### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by Newtayo**
## Hello Reviewer
@danielmatama  and I have completed the recipe app project. Below are some of the features we implemented
## 

- Login page and registration page built with devise
- A Food list that displays a list of food added by the logged-in user as in the wireframe (display also quantity of a given food). it also leads to a form that allows users to add new food.
- Recipes list: displays a list of recipes created by the logged-in user as in the wireframe. It leads to recipe details. If the user is the owner of the recipe, should allow the user to delete it.
- Public recipe list: it displays a list of all public recipes ordered by newest as in the wireframe. it leadsto recipe details.
- Recipe details: has a toggle button that allows for a recipe to be made public or private. If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe. If the user is the owner of the recipe, should lead to the form that allows the user to add new food. Should count the total food items and total price of the missing food.
Make sure there are no N+1 queries happening.
Create a navigation menu that allows users to open all of the pages you created.
Write unit and integration tests

### CODE REVIEW

File: app/views/shopping_lists/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        <% unless food[:price].zero? %>
+          <tr>
+            <th><%= food[:food].name %></th>
+            <th><%= food[:quantity] %><%= food[:food].measurement_unit %></th>
+            <th>$ <%= food[:price] %></th>
+          </tr>
+        <% end %>
+      <% end %>
+    </tbody>
+    </table>
```
Comment (danmainah) [inline app/views/shopping_lists/index.html.erb:25]: - [ ] @Newtayo and @danielmatama Great work adding a  shopping list page, however, am unable to generate a shopping list from my recipe list

`mylist`
image

`shopping list page after I click generate shopping list button`
image

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    can :read, User
+    can :manage, User
+    can :manage, RecipeFood
+    return unless user.present?
+
+    can :read, Recipe, public: true
+    can :manage, Recipe, user_id: user.id
+    can :read, Food
+    can :manage, Food, user_id: user.id
+  end
```
Comment (danmainah) [inline app/models/ability.rb:14]: - [ ] @team Great work adding the ability model to restrict access however you didn't add the functionality to restrict only the owner of the to have the ability to add new food to the recipe ,currently all users can access the add ingredients page also can delete ingredients

File: app/controllers/foods_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def destroy
+    @food = Food.find(params[:id])
+    @recipe_food = @food.recipe_foods.where(food_id: @food.id)
+    @recipe_food.each(&:destroy)
+    @food.destroy
+    respond_to do |format|
+      format.html { redirect_to foods_url, notice: 'Food was successfully destroyed.' }
+    end
```
Comment (danmainah) [inline app/controllers/foods_controller.rb:31]: - [ ] @team great work adding the destroy method however you forgot to add a fail path incase the food is not destroyed.

File: app/controllers/recipe_foods_controller.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  def recipe_food_params
+    params.require(:recipe_food).permit(:quantity, :food_id)
+  end
+
+  def destroy
+    @user = current_user
+    @recipe = Recipe.find(params[:recipe_id])
+    @recipe_food = @recipe.recipe_foods.find(params[:id])
+    @recipe_food.destroy
```
Comment (danmainah) [inline app/controllers/recipe_foods_controller.rb:51]: - [ ] @team Great work adding the destroy method however you forgot to add the success and fail path to alert the user if the recipe_food is destroyed or not.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Newtayo and @danielmatama , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Good design, the design of the pages matches the wireframe given
✔️ Usera are able to signup/login successfully
✔️ Users can add recipe_list, food, and recipe items and make recipe_list public or remove it from public view.


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/views/shopping_lists/index.html.erb (hunk 1 above)
Comment (Newtayo) [reply app/views/shopping_lists/index.html.erb:25]: @danmainah Thank you for the feedback. However, the general shopping list will only generate a list if the quantity of food you need for your recipe is less than the quantity you have in your food list

**Review by sumairq — APPROVED** [review]
### Hi  Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
✔️ Good commit history
✔️Linters are passing
✔️Very professional README
✔️Correct Git flow
✔️Descriptive PR
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Newtayo** [comment]
@sumairq Thank you

**Pull request merged**